### DISCUSSION TITLE
Joins CLI without flatten

### DISCUSSION THREAD

**Discussion by dqnhat** [donmccurdy/glTF-Transform#1272] [Q&A] [answered]
Hi there, i'm using @gltft-transform/cli install from nodejs for some backend optimize work.

I want to "Joins compatible Primitives" which "are members of the same Mesh" while "Prevents joining distinct Meshes and Nodes"
> // JOIN
program.command('join', 'Join meshes and reduce draw calls').help(`Joins compatible Primitives and reduces draw calls. Primitives are eligible for joining if they are members of the same Mesh or, optionally, attached to sibling Nodes in the scene hierarchy. Implicitly runs \`dedup\` and \`flatten\` commands first, to maximize the number of Primitives that can be joined.
NOTE: In a Scene that heavily reuses the same Mesh data, joining may increase vertex count. Consider alternatives, like \`instance\` with EXT_mesh_gpu_instancing.)

Currently i'm using
`gltf-transform join ori.gltf dest.gltf --keepMeshes true --keepNamed true`
but the node structure keep getting change, which is because in the `join` command there is a `flatten` command

> // FLATTEN
program.command('flatten', 'Flatten scene graph').help(` Flattens the scene graph, leaving Nodes with Meshes, Cameras, and other attachments as direct children of the Scene. Skeletons and their descendants are left in their original Node structure.)

currently i can commenting out the `flatten` part inside the code "cli.esm.js" but that is a terrible idea long run.
`optimize` also have too must restrict that i cannot use.
![image](https://github.com/donmccurdy/glTF-Transform/assets/139830344/9a059e1d-b425-4e3b-a93e-90a71a943e93)

i want to ask for some thing like `gltf-transform join ori.gltf dest.gltf --keepMeshes true --keepNamed true --flatten false` or a way to using `join`  directly without `flatten`. Basically [donmccurdy/glTF-Transform/discussions/1009] but for `join` and `flatten`
or if i just not understand how this work.

Any other suggestion, recommendation, criticize are welcome.

**Comment by donmccurdy** [comment, marked as answer]
Hi @dqnhat — currently the CLI does not offer the level of control you're looking for here. If you open the model in https://gltf.report/, open the script tab, and paste and run the script below, does that give the result you expect?

```js
import { join } from '@gltf-transform/functions';

await document.transform(join());
```
